staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,297 @@
+	// storageversion completes. This is used by the storageversion manager
+	// to unblock publishing of the latest storageversion.
+	teardownFinishedCh <-chan struct{}
+}
+
+// NewManager creates a CRD StorageVersion Manager.
+func NewManager(client genericstorageversion.Client, apiserverID string) Manager {
+	return Manager{
+		client:      client,
+		apiserverID: apiserverID,
+		queue:       workqueue.NewNamed("storageversion-updater"),
+	}
+}
+
+func (m *Manager) Enqueue(crd *apiextensionsv1.CustomResourceDefinition, tearDownFinishedCh <-chan struct{}) {
+	m.recordLatestSVUpdateInfoWithTeardown(crd, tearDownFinishedCh)
+	m.queue.Add(crd.UID)
+}
+
+func (m *Manager) WaitForStorageVersionUpdate(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+	var latestSVUpdateInfo *storageVersionUpdateInfo
+	val, found := latestStorageVersionUpdateInfo.Load(crd.UID)
+	if !found {
+		return nil
+	}
+
+	latestSVUpdateInfo = val.(*storageVersionUpdateInfo)
+
+	// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+	// to register themselves to the wait group. Ideally the storage version update should
+	// not cause the requests to miss the 1s window; otherwise the requests may
+	// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+	// first CR request establishes the underlying storage).
+	select {
+	case <-latestSVUpdateInfo.updateChannels.errCh:
+		return fmt.Errorf("error while waiting for CRD storage version update")
+	case <-latestSVUpdateInfo.updateChannels.processedCh:
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("aborted waiting for CRD storage version update: %w", ctx.Err())
+	// Unblock the requests if the storage version update takes a long time, otherwise
+	// CR requests may stack up and overwhelm the API server.
+	// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+	case <-time.After(storageVersionUpdateTimeout):
+		return fmt.Errorf("timeout waiting for CRD storage version update")
+	}
+
+}
+
+func (m *Manager) RunUpdateLoop(stopCh <-chan struct{}) {
+	go wait.Until(m.worker, time.Second, stopCh)
+	defer m.queue.ShutDown()
+	<-stopCh
+}
+
+func (m *Manager) worker() {
+	for m.processLatestSVUpdate() {
+	}
+}
+
+func (m *Manager) processLatestSVUpdate() bool {
+	ctx := context.TODO()
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+	defer m.queue.Done(key)
+
+	val, ok := latestStorageVersionUpdateInfo.Load(key)
+	latestSVUpdateInfo := val.(*storageVersionUpdateInfo)
+	if !ok || latestSVUpdateInfo.processed {
+		klog.V(2).Infof("No pending storageversion update found, returning")
+		return true
+	}
+
+	m.updateStorageVersion(ctx, latestSVUpdateInfo.crd, latestSVUpdateInfo.updateChannels.teardownFinishedCh, latestSVUpdateInfo.updateChannels.processedCh, latestSVUpdateInfo.updateChannels.errCh)
+	markUpdateAsProcessed(latestSVUpdateInfo)
+	return true
+}
```

Comment:
Should we re-enqueue? I'm unclear on how the SV update is coordinated when the SV update request times out.

Loosely related:  If the CRD update succeeds, and the cluster has only 1 apiserver, and that apiserver crashes after the CRD update but before the SV update, what happens?

Reply:
> Should we re-enqueue?

That might be a good suggestion. We do retries on just the storageversion update [here](https://github.com/kubernetes/kubernetes/blob/release-1.29/staging/src/k8s.io/apiserver/pkg/storageversion/updater.go#L126-L146), but there can be failures outside of this function. Like if the teardown of the old storage times out for any reason, do we want try the SV update (along with the teardown) again? Open to suggestions. If we re-enqueue, then for how long after the first failure do we keep doing that?

Reply:
> If the CRD update succeeds, and the cluster has only 1 apiserver, and that apiserver crashes after the CRD update but before the SV update, what happens?

From my understanding, the old handler and old storage would be torn down. When the apiserver comes up again:
1. new handler (using new storageversion) is created when a CR request arrives
2. the new CR request triggers waitForStorageVersionUpdate() which tries to fetch the latest entry from the storageVersionUpdateInfoMap
3. finds nothing there, and returns nil
4. CR is served using new storage but the storageversion API doesnt reflect that. Storageversion migrator wouldn't trigger the migration, unless a new CRD update happens.

Maybe we should change the behavior at #3. We can invoke the SV Update here instead of returning a nil.
cc @roycaihw for verifying my understanding and his thoughts on the same.

Reply:
If we do the above, it also means that whenever a CR write arrives after the server is restarted, we will queue the relevant CRD for SV update regardless of whether its SV was already updated (i.e. before the server crashed)

Reply:
When a server reboots, it reconciles all the CRDs and update SVs accordingly. New CR write requests should be blocked until the corresponding SV is up-to-date.

Reply:
I added [requeuing](https://github.com/kubernetes/kubernetes/pull/123999/files#diff-ad10c66dede77a3da8db2294a745acdef0fbdd4f7e7ac8d18b835572572943a3R128-R138) for whenever a CRD is not found in the storageversionUpdateInfoMap to accommodate for server bootstrap

Reply:
Note: The requeuing would only happen when a CR write is observed after the server restarts. Which means that the SV for a CRD would not be reconciled **until** a corresponding CR write is made. This doesn't seem very ideal to me. Maybe we should instead have a poststarthook that does a list on all available CRDs and queues the SV updates for all CRDs found in the list? @roycaihw @jpbetz

Reply:
I have modified the [existing poststarthook](https://github.com/kubernetes/kubernetes/blob/release-1.29/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L268-L276) (which currently only waited for the crd informer to sync, to also wait till all crds are enqueued for SV updates).